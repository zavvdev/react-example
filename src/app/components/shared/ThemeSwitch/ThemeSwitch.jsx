import cx from "clsx";
import PropTypes from "prop-types";
import { useThemeSwitchStyles } from "app/components/shared/ThemeSwitch/ThemeSwitch.styles";

export function ThemeSwitch({ isDark, onToggle, className }) {
  const classes = useThemeSwitchStyles();
  const rootClasses = cx(classes.root, className);

  const switcherClasses = cx(classes.switcher, {
    [classes.switcherDark]: isDark,
  });

  return (
    <button type="button" className={rootClasses} onClick={onToggle}>
      <div className={switcherClasses}>{isDark ? "üåö" : "üåù"}</div>
    </button>
  );
}

ThemeSwitch.propTypes = {
  isDark: PropTypes.bool.isRequired,
  onToggle: PropTypes.func.isRequired,
  className: PropTypes.string,
};

ThemeSwitch.defaultProps = {
  className: undefined,
};
