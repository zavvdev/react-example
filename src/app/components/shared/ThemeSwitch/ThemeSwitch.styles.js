import { createUseStyles } from "react-jss";

export const useThemeSwitchStyles = createUseStyles((theme) => ({
  root: {
    width: "40px",
    height: "20px",
    backgroundColor: theme.components.ThemeSwitch.trackBackgroundColor,
    borderRadius: "20px",
    position: "relative",
    border: "none",
    cursor: "pointer",
  },
  switcher: {
    display: "flex",
    justifyContent: "center",
    alignItems: "center",
    position: "absolute",
    top: "50%",
    transform: "translate(0, -50%)",
    left: -1,
    fontSize: "30px",
  },
  switcherDark: {
    right: -1,
    left: "unset",
  },
}));
