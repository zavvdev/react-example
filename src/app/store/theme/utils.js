import { localStorageService } from "app/services/LocalStorageService";
import { IS_DARK_THEME_LOCAL_STORAGE_KEY } from "app/store/theme/config";

export const saveDarkModeState = (darkModeState) => {
  localStorageService.set(IS_DARK_THEME_LOCAL_STORAGE_KEY, `${darkModeState}`);
};

export const getDarkModeState = () => {
  const rawStorageValue = localStorageService.get(
    IS_DARK_THEME_LOCAL_STORAGE_KEY,
  );
  return !!JSON.parse(rawStorageValue);
};
