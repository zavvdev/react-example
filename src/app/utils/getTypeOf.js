export function getTypeOf(object) {
  const objectType = typeof object;
  let returnType = objectType;
  if (objectType === "object") {
    const stringifiedObject = Object.prototype.toString.call(object);
    returnType = stringifiedObject
      .replace(/^\[object (\S+)\]$/, "$1")
      .toLowerCase();
  }
  return returnType;
}
