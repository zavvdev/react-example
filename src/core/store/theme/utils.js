import { LOCAL_STORAGE_KEYS } from "core/config/localStorage";
import { localStorageService } from "core/services/LocalStorageService";

export const saveDarkModeState = (darkModeState) => {
  localStorageService.set(
    LOCAL_STORAGE_KEYS.isThemeDarkMode,
    `${darkModeState}`,
  );
};

export const getDarkModeState = () => {
  const rawStorageValue = localStorageService.get(
    LOCAL_STORAGE_KEYS.isThemeDarkMode,
  );
  return !!JSON.parse(rawStorageValue);
};
