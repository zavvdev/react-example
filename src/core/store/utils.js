import { all, call, spawn } from "redux-saga/effects";
import { errorTrackingService } from "core/services/ErrorTrackingService";

export const combineMiddleware = (middlewareList) => function* combiner() {
  const getRootGenerator = (middleware) => function* rootGenerator() {
    while (true) {
      try {
        yield call(middleware);
        break;
      } catch (e) {
        errorTrackingService.reportError(e);
      }
    }
  };

  yield all(middlewareList.map((middleware) => {
    return spawn(getRootGenerator(middleware));
  }));
};
