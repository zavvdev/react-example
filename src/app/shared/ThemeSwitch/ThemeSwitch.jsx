import cx from "clsx";
import PropTypes from "prop-types";
import { useThemeSwitchStyles } from "app/shared/ThemeSwitch/ThemeSwitch.styles";
import { useAppTranslation } from "app/i18n/useAppTranslation";

export function ThemeSwitch({ isDark, onToggle, className }) {
  const { t } = useAppTranslation();
  const classes = useThemeSwitchStyles();
  const rootClasses = cx(classes.root, className);

  const switcherClasses = cx(classes.switcher, {
    [classes.switcherDark]: isDark,
  });

  return (
    <button type="button" className={rootClasses} onClick={onToggle}>
      <div className={switcherClasses}>
        {isDark ? t("themeSwitch.dark") : t("themeSwitch.light")}
      </div>
    </button>
  );
}

ThemeSwitch.propTypes = {
  isDark: PropTypes.bool.isRequired,
  onToggle: PropTypes.func.isRequired,
  className: PropTypes.string,
};

ThemeSwitch.defaultProps = {
  className: undefined,
};
