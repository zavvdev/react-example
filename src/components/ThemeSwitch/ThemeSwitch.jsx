import { MoonIcon } from "components/svgIcons/MoonIcon";
// import { SunIcon } from "components/svgIcons/SunIcon";
import { useThemeSwitchStyles } from "components/ThemeSwitch/ThemeSwitch.styles";

function ThemeSwitch() {
  const classes = useThemeSwitchStyles();

  return (
    <button type="button" className={classes.root}>
      <div className={classes.switcher}>
        <MoonIcon />
      </div>
      {/* <div>
        <SunIcon />
      </div>
      <div>
        <MoonIcon />
      </div> */}
    </button>
  );
}

export { ThemeSwitch };
