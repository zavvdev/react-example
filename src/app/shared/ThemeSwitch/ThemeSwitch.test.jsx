import { fireEvent, screen } from "@testing-library/react";
import { ThemeSwitch } from "app/shared/ThemeSwitch/ThemeSwitch";
import { testUtils } from "__tests__/utils";

const properties = {
  onToggle: jest.fn(),
  className: "mock",
};

describe("ThemeSwitch", () => {
  test("should render ThemeSwitch (dark mode)", () => {
    const result = testUtils.render(
      <ThemeSwitch isDark onToggle={properties.onToggle} />,
    );
    expect(result.asFragment()).toMatchSnapshot();
  });

  test("should render ThemeSwitch (light mode)", () => {
    const result = testUtils.render(
      <ThemeSwitch isDark={false} onToggle={properties.onToggle} />,
    );
    expect(result.asFragment()).toMatchSnapshot();
  });

  test("should have a custom className", () => {
    const result = testUtils.render(
      <ThemeSwitch
        isDark
        onToggle={properties.onToggle}
        className={properties.className}
      />,
    );
    expect(
      result.container.querySelectorAll(`.${properties.className}`).length,
    ).toBe(1);
  });

  test("should call onToggle function", () => {
    testUtils.render(<ThemeSwitch isDark onToggle={properties.onToggle} />);
    fireEvent.click(screen.getByRole("button"));
    expect(properties.onToggle).toBeCalled();
  });
});
