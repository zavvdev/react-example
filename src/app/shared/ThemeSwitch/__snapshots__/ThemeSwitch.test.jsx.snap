// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemeSwitch should render ThemeSwitch (dark mode) 1`] = `
<DocumentFragment>
  <button
    class="root"
    type="button"
  >
    <div
      class="switcher switcherDark"
    >
      themeSwitch.dark
    </div>
  </button>
</DocumentFragment>
`;

exports[`ThemeSwitch should render ThemeSwitch (light mode) 1`] = `
<DocumentFragment>
  <button
    class="root"
    type="button"
  >
    <div
      class="switcher"
    >
      themeSwitch.light
    </div>
  </button>
</DocumentFragment>
`;
